vms/evm/subscriber.go

```diff
@@ -182,17 +206,66 @@ func (s *subscriber) subscribe(retryTimeout time.Duration) error {
 		return err
 	}
 	s.sub = sub
+
+	go s.blocksInfoFromHeaders()
 	return nil
 }
 
-func (s *subscriber) Headers() <-chan *types.Header {
-	return s.headers
+// blocksInfoFromHeaders listens to the header channel and converts the headers to [relayerTypes.WarpBlockInfo]
+// and writes them to the blocks channel consumed by the listener
+func (s *subscriber) blocksInfoFromHeaders() {
+	for header := range s.headers {
+		block, err := relayerTypes.NewWarpBlockInfo(s.logger, header, s.rpcClient)
+		if err != nil {
+			s.logger.Error("Failed to create Warp block info", zap.Error(err))
+			s.errChan <- err
+			return
+		}
+		s.icmBlocks <- block
+	}
 }
 
-func (s *subscriber) Err() <-chan error {
+func (s *subscriber) ICMBlocks() <-chan *relayerTypes.WarpBlockInfo {
+	return s.icmBlocks
+}
+
+// SubscribeErr returns the error channel for the underlying subscription
+func (s *subscriber) SubscribeErr() <-chan error {
 	return s.sub.Err()
 }
 
+// Err returns the error channel for miscellaneous errors not recoverable from
+// by resubscribing.
+func (s *subscriber) Err() <-chan error {
+	return s.errChan
+}
+
 func (s *subscriber) Cancel() {
 	// Nothing to do here, the ethclient manages both the log and err channels
 }
+
+// Takes a block range and a slice of blocks containing ICM messages and fills in the missing blocks with empty blocks
+func fillBlockRange(fromBlock, toBlock uint64, icmBlocks []*relayerTypes.WarpBlockInfo) ([]*relayerTypes.WarpBlockInfo, error) {
+	var filledBlocks []*relayerTypes.WarpBlockInfo
+	currentBlock := fromBlock
+	for _, block := range icmBlocks {
```

Comment:
It would be slightly less efficient (actually, maybe not because of all the appending we do), but much more straightforward if we just created an array of empty blocks for the whole range, and then overwrote the entries with the `icmBlocks`